:one::three::tada: Init commit

fedoranvar opened the pull request.

yelizariev commented on base_api/README.rst:
```diff
@@ -0,0 +1,51 @@
+
+basic function and methods of API for openapi or XML-RPC
+
+Models and controllers for basic functionality
```
В модуле не планируется контроллеры

yelizariev commented on base_api/README.rst:
```diff
@@ -0,0 +1,51 @@
+ Base Api
+==========
+
+basic function and methods of API for openapi or XML-RPC
```
```suggestion
Basic methods of API for openapi or XML-RPC
```

yelizariev commented on base_api/__manifest__.py:
```diff
@@ -0,0 +1,48 @@
+
+{
+    "name": """Base API""",
+    "summary": """basic function and methods of API for openapi or XML-RPC""",
```
С большой буквы

yelizariev commented on base_api/doc/index.rst:
```diff
@@ -0,0 +1,35 @@
+Uninstallation
+==============
+
+{Optional section for uninstallation notes. Delete it if you don't have notes for uninstallation.}
```
![удоли](https://user-images.githubusercontent.com/186131/69955891-86078d00-1520-11ea-994a-4201eb6f9fcf.jpg)

yelizariev commented on base_api/README.rst:
```diff
@@ -0,0 +1,167 @@
+Functionality
+=============
+
+**base.py** - extends *base*-model with functions like:
```
It's better to say "adds following methods to all models"

yelizariev commented on base_api/README.rst:
```diff
@@ -0,0 +1,167 @@
+
+* **search_or_create**(self, vals, active_test=True)
+    *In this function "race conditions" resolved, so there won't be record-duplication*
+    *It checks if there are records that satisfies required `vals`, if there is, then it return they* `ids` *else creates one and returns it id*
```
fix formatting
![image](https://user-images.githubusercontent.com/186131/70240392-f7e50e00-178e-11ea-89e5-6f14ab3b96fc.png)

yelizariev commented on base_api/README.rst:
```diff
@@ -0,0 +1,167 @@
+* **search_or_create**(self, vals, active_test=True)
+    *In this function "race conditions" resolved, so there won't be record-duplication*
+    *It checks if there are records that satisfies required `vals`, if there is, then it return they* `ids` *else creates one and returns it id*
+    **note**: for x2many and many2one fields creation didn't work 
```
x2many: Создание работает. Поиск не работает. 

many2one: работает для айди передавать

yelizariev replied:
И надо писать что означает "не работает"

yelizariev commented on base_api/README.rst:
```diff
@@ -0,0 +1,167 @@
+    *It checks if there are records that satisfies required `vals`, if there is, then it return they* `ids` *else creates one and returns it id*
+    **note**: for x2many and many2one fields creation didn't work 
+
+    * searches for records with selected fields with values (``vals``)
```
Про это уже написано выше

yelizariev commented on base_api/__init__.py:
```diff
@@ -0,0 +1,7 @@
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html)
+
+def post_load():
+    # make import in post_load to avoid applying monkey patches when this
```
There are no monkey patches in this module

yelizariev commented on base_api/controllers/pinguin.py:
```diff
@@ -0,0 +1,311 @@
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
```
No need to put this in in controllers folder

yelizariev commented on base_api/controllers/pinguin.py:
```diff
@@ -0,0 +1,311 @@
+            elif (value is False or value is None) and fld.type != 'boolean':
+                # string field cannot be false in response json
+                result[field] = ''
+    return result
```
You must delete copy-pasted code from openapi module

yelizariev commented on base_api/doc/index.rst:
```diff
@@ -0,0 +1,8 @@
+Usage
+=====
+
+Use it with such models as like as OpenAPI
```
OpenAPI is not a model.
Don't forget about buit-in XML RPC

yelizariev commented on base_api/tests/test_base.py:
```diff
@@ -0,0 +1,226 @@
+        # Test #1: Record creation
+        #
+        t_vals = {'name': t_name}
+        partner_obj.search([('name', '=', t_name)]).unlink()
```
You delete records here, but don't it with next tests

            'name': 'TestParent',


Why?

yelizariev reviewed (commented):
Добавь base_api   в зависимости openapi

yelizariev reviewed (changes requested):
Надо описание функций перенести в index.rst всё таки. Иначе в магазине пусто будет

yelizariev commented on base_api/doc/index.rst:
```diff
@@ -0,0 +1,9 @@
+=====
+
+* As a dependency in OpenAPI module
+* Can be used in built-in XML RPC
```
Вот что надо написать: 

_методы предназначенные для вызова извне через апи (например через Опеапи или РПЦ)_

Далее сделать заголовок Методы и перенести описание методов из РИДМИ

yelizariev commented on base_api/controllers/pinguin.py:
```diff
@@ -0,0 +1,310 @@
+# pyling: disable=redefined-builtin
```
Верни лицензию

yelizariev commented on openapi/doc/index.rst:
```diff
@@ -10,6 +10,8 @@ Installation
 
 * `Install <https://odoo-development.readthedocs.io/en/latest/odoo/usage/install-module.html>`__ this module in a usual way
 * Add ``openapi`` to `--load parameter <https://odoo-development.readthedocs.io/en/latest/admin/server_wide_modules.html>`__, e.g.::
+  + Note: Skipping of this step may lead to ignorance of a first request by
```
Проверяй форматирование

![screenshot-github com-2019 12 19-15_07_33](https://user-images.githubusercontent.com/186131/71164643-50c5a380-2271-11ea-82f3-45ddba8a48fa.png)

yelizariev replied:
Ты не написал какая именно ошибка

yelizariev commented on base_api/static/description/index.html:
```diff
@@ -0,0 +1,3 @@
+<div>
+    please add some information
```
:-1:

yelizariev replied:
Don't cheat check for documentation

yelizariev commented on base_api/doc/index.rst:
```diff
@@ -0,0 +1,261 @@
+      - it returns list of record-dictionaries with nested fields
+        from related models (via *x2many*, *many2one*), by using
+        dot-notation in `fields`
+      - e.g. `fields = [ company_id, company_id.name ]`
```
Еще отличие в новом аргументе delimeter

yelizariev replied:
Либо сделай точку по умолчанию для разделителя, либо примеры поменяй на слеш

yelizariev commented on base_api/models/base.py:
```diff
@@ -0,0 +1,79 @@
+                        elif list_record[0] == 6:
+                            for record_for_replace in list_record[2]:
+                                if isinstance(record_for_replace, str):
+                                    record_for_replace = ird_object.xmlid_lookup(id_prefix + '.' + record_for_replace)[2]
```
Слишком много вложений. Переделай как-нибудь

fedoranvar replied:
разделил на два цикла

yelizariev commented on base_api/models/base.py:
```diff
@@ -0,0 +1,79 @@
+
+    @api.model
+    def create_or_update_by_external_id(self, vals):
+        id_prefix = '__base_api__'
```
consts must be at the beginning of the file

yelizariev commented on base_api/models/base.py:
```diff
@@ -0,0 +1,79 @@
+            ird_object.create({
+                'name': vals.get('id'),
+                'model': self._name,
+                'module': '__base_api__',
```
you have to use the const here

yelizariev commented on base_api/lib/pinguin.py:
```diff
@@ -0,0 +1,318 @@
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
```
Добавь свои и оригинальные копирайты

yelizariev commented on base_api/doc/index.rst:
```diff
@@ -0,0 +1,261 @@
+      - Must contain `id` field type of *string*
+      - e.g. `{ 'id': 'ext.id_1', 'name: 'John', 'age': 37,
+        'job_id': 'ext.work_1', 'child_ids' : [(4, 'ext.child_1', 0),
+        (4, 'ext.child_2', 0)] }`
```
Для кода делай соответствующее форматирование

![screenshot-github com-2019 12 19-15_34_23](https://user-images.githubusercontent.com/186131/71166694-16f69c00-2275-11ea-9aa0-4692d626c765.png)

fedoranvar replied:
чек

yelizariev commented on base_api/models/base.py:
```diff
@@ -0,0 +1,79 @@
+                                    record_for_replace = ird_object.xmlid_lookup(id_prefix + '.' + record_for_replace)[2]
+                        vals[field][index] = tuple(list_record)
+            except ValueError:
+                raise ValueError('No external id for %s' % (field))
```
Непонятно что это значит. Не записи с таким айди или вместо айди-строки получили числи или вообще там пусто?

yelizariev replied:
Если речь про то, что по айди-строке не нашли запись, то желательно писать чему равна эта строка -- в одном поле может несколько строк быть

fedoranvar replied:
речь о том, что по айди-строке не нашли. Чтобы не плодить вложенность, я оставлю только один try, но переделаю сообще об ошибке

yelizariev commented on base_api/models/base.py:
```diff
@@ -0,0 +1,79 @@
+        ird_object = self.env['ir.model.data']
+        # if external id not defined
+        if not isinstance(ext_id, str):
+            raise ValueError('External ID is not defined')
```
Описание ошибки не очень. У тебя в документации про это по другому совсем написано:


      - Must contain `id` field type of *string*

yelizariev reviewed (changes requested):
* I'm not sure that the icon is correct

yelizariev commented on base_api/models/base.py:
```diff
@@ -0,0 +1,79 @@
+                'module': '__base_api__',
+                'res_id': inner_id
+                })
+        # Yes: Write changes to record
```
I'd prefer to put it inside the block

yelizariev commented on openapi/security/ir.model.access.csv:
```diff
@@ -2,7 +2,7 @@ id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
 user_access_openapi_namespace,access_openapi_namespace,model_openapi_namespace,openapi.group_user,1,0,0,0
 user_access_openapi_access,access_openapi_access,model_openapi_access,openapi.group_user,1,0,0,0
-user_access_openapi_access_create_context,access_openapi_access_create_context,model_openapi_access_create_context,openapi.group_user,1,0,0,0
+user_access_openapi_access_create_context,access_openapi_access_create_context,model_openapi_access_create_context,openapi.group_user,1,1,1,0
```
https://github.com/it-projects-llc/sync-addons/pull/69#discussion_r361938806

Commit:
:one::three::tada: init commit
:euro: decreased due to creation of base_api
:book: description of error by skipping step about 'server_wide_modules'
:rainbow: refactoring due to creation of 'base_api' module

yelizariev commented:
Этот комит надо объединить с комитов создания модуля

> rainbow refactoring due to creation of 'base_api' module

Там никакой не рефакторинг, а удаление кода

yelizariev commented:
Да и изменение цены отдельно не желательно делать. Это всё часть одного большого изменения, которое нужно вместе принимать, вместе переносить в другие ветки и т.д.

yelizariev commented:
# [Codecov](https://codecov.io/gh/itpp-labs/sync-addons/pull/64?src=pr&el=h1) Report
> Merging [#64](https://codecov.io/gh/itpp-labs/sync-addons/pull/64?src=pr&el=desc) into [13.0](https://codecov.io/gh/itpp-labs/sync-addons/commit/e79057133d8cd548f4fe1e85db06e80959d31b3b&el=desc) will **increase** coverage by `3.64%`.
> The diff coverage is `n/a`.

[![Impacted file tree graph](https://codecov.io/gh/itpp-labs/sync-addons/pull/64/graphs/tree.svg?width=650&height=150&src=pr&token=GwjhtrwSXC)](https://codecov.io/gh/itpp-labs/sync-addons/pull/64?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##             13.0      #64      +/-   ##
==========================================
+ Coverage   74.32%   77.97%   +3.64%     
==========================================
  Files          16       17       +1     
  Lines        1071     1212     +141     
==========================================
+ Hits          796      945     +149     
+ Misses        275      267       -8     
```

| Flag | Coverage Δ | |
|---|---|---|
| #backend | `77.97% <91.31%> (+3.64%)` | :arrow_up: |
| #controllers | `77.97% <91.31%> (+3.64%)` | :arrow_up: |
| #res_config | `77.97% <91.31%> (+3.64%)` | :arrow_up: |
| #tests | `77.97% <91.31%> (+3.64%)` | :arrow_up: |

| [Impacted Files](https://codecov.io/gh/itpp-labs/sync-addons/pull/64?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [base\_api/tests/test\_base.py](https://codecov.io/gh/itpp-labs/sync-addons/pull/64/diff?src=pr&el=tree#diff-YmFzZV9hcGkvdGVzdHMvdGVzdF9iYXNlLnB5) | `100.00% <0.00%> (ø)` | |
| [base\_api/models/base.py](https://codecov.io/gh/itpp-labs/sync-addons/pull/64/diff?src=pr&el=tree#diff-YmFzZV9hcGkvbW9kZWxzL2Jhc2UucHk=) | `88.88% <0.00%> (ø)` | |
| [base\_api/lib/pinguin.py](https://codecov.io/gh/itpp-labs/sync-addons/pull/64/diff?src=pr&el=tree#diff-YmFzZV9hcGkvbGliL3Bpbmd1aW4ucHk=) | `85.03% <0.00%> (ø)` | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/itpp-labs/sync-addons/pull/64?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/itpp-labs/sync-addons/pull/64?src=pr&el=footer). Last update [e790571...8665077](https://codecov.io/gh/itpp-labs/sync-addons/pull/64?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

yelizariev reviewed (approved):
Looks good
